packages/block-library/src/image/edit.js

```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
 	}
 
 	updateAlignment( nextAlign ) {
-		const extraUpdatedAttributes = [ 'wide', 'full' ].indexOf( nextAlign ) !== -1 ?
-			{ width: undefined, height: undefined } :
-			{};
-		this.props.setAttributes( { ...extraUpdatedAttributes, align: nextAlign } );
+		if ( nextAlign === 'wide' || nextAlign === 'full' ) {
+			// Reset all sizing attributes.
+			this.resetWidthHeight();
+		}
+
+		this.props.setAttributes( { align: nextAlign } );
+	}
+
+	updateImageURL( url, imageSizeOptions ) {
+		this.resetWidthHeight();
+		let fileWidth;
+		let fileHeight;
+
+		if ( imageSizeOptions.width && imageSizeOptions.height ) {
+			fileWidth = imageSizeOptions.width;
+			fileHeight = imageSizeOptions.height;
+		} else {
+			// Find the image data.
+			map( imageSizeOptions, ( { value, imageData } ) => {
+				if ( value === url ) {
+					fileWidth = imageData.width;
+					fileHeight = imageData.height;
+					return false;
+				}
+			} );
+		}
+
+		this.props.setAttributes( {
+			url,
+			fileWidth,
+			fileHeight,
+		} );
+	}
+
+	updateWidth( width, fileWidth, fileHeight, userSet ) {
+		width = parseInt( width, 10 );
+
+		// Reset the image size when the user deletes the value.
+		if ( ! width || ! fileWidth || ! fileHeight ) {
+			this.resetWidthHeight();
+			return;
+		}
+
+		const height = round( fileHeight * ( width / fileWidth ) );
+		this.setWidthHeight( width, height, fileWidth, fileHeight, userSet );
+	}
+
+	updateHeight( height, fileWidth, fileHeight, userSet ) {
+		height = parseInt( height, 10 );
+
+		// Reset the image size when the user deletes the value.
+		if ( ! height || ! fileWidth || ! fileHeight ) {
+			this.resetWidthHeight();
+			return;
+		}
+
+		const width = round( fileWidth * ( height / fileHeight ) );
+		this.setWidthHeight( width, height, fileWidth, fileHeight, userSet );
 	}
 
-	updateImageURL( url ) {
-		this.props.setAttributes( { url, width: undefined, height: undefined } );
+	setWidthHeight( width, height, fileWidth, fileHeight, userSet ) {
+		// Set image size and also whether set directly by the user.
+		userSet = userSet || undefined;
+
+		this.props.setAttributes( {
+			width,
+			height,
+			fileWidth,
+			fileHeight,
+			userSet,
+		} );
 	}
 
-	updateWidth( width ) {
-		this.props.setAttributes( { width: parseInt( width, 10 ) } );
+	resetWidthHeight( fileWidth, fileHeight ) {
+		if ( fileWidth && fileHeight ) {
+			this.props.setAttributes( {
+				fileWidth,
+				fileHeight,
+			} );
+		}
+
+		this.props.setAttributes( {
+			width: undefined,
+			height: undefined,
+			userSet: undefined,
+		} );
```

Comment:
there's no need here for two calls to `setAttributes()` - we can collapse them into one in a couple of different ways. my favorite is with `Object.assign()`

```js
// what does this name mean?
resetWidthHeight( fileWidth, fileHeight ) {
	this.props.setAttributes(
		{
			width: undefined,
			height: undefined,
			userSet: undefined,
		},
		fileWidth && fileHeight && {
			fileWidth,
			fileHeight,
		}
	);
}
```

Reply:
Sure. The purpose here is to make it as easy to read and understand as possible. After going through babel and minification, the code that actually runs in production will be completely different.

(I actually "have a beef" with the way a lot of the code is written. All the redundant indentation, nested lambda functions (no names = poor context), "clever" conditionals, almost no inline comments... There seem to be quite a few cases where the code becomes more readable after running it through bable and minification, then "unminifying" it. But lets look at that after 5.0 is out.) :)

Reply:
> there's no need here for two calls to `setAttributes()` - we can collapse them into one in a couple of different ways. my favorite is with `Object.assign()`

(Noting that the supplied example doesn't use `Object.assign` and wouldn't work as written)

Reply:
> (Noting that the supplied example doesn't use Object.assign and wouldn't work as written)

Good call! 😆

Reply:
> Sure. The purpose here is to make it as easy to read and understand as possible. After going through babel and minification, the code that actually runs in production will be completely different.

The main thing for me here would be to avoid multiple calls to `setAttributes`, as it will incur a render for each (unnecessary impact on performance). This will remain true after transpilation and minification.

I pushed bf67408 which I think is a reasonable compromise to preserve a similar readability.